Leader does not respond to ReadIndex message from learner

In a cluster with 1 leader and 1 learner (quorum == 1), a linearizable read request sent to learner node will timeout. More detailed logs can be found [in this issue](https://github.com/jingyih/etcd/issues/7). Basically, the learner node will timeout after sending out ReadIndex message to leader and waiting for leader's response.

**What is supposed to happen**

1. A linearizable read request notifies the `LinearizableReadLoop`.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/etcdserver/v3_server.go#L94-L96

2. Upon being notified, `LinearizableReadLoop` sends a ReadIndex request to local raft node. In this case, this node is a raft learner.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/etcdserver/v3_server.go#L651

3. The learner node, also as a follower, upon receiving ReadIndex message, forwards the message to leader.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/raft/raft.go#L1291-L1297

4. Leader receives the message, sends out heartbeat to ensure it is not deposed by a more recent leader (for simplicity let's only talk about `ReadOnlySafe` option)
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/raft/raft.go#L1025-L1039

5. Eventually leader sends out ReadIndex response after all safety checks are passed.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/raft/raft.go#L1142


**What went wrong**
In a cluster with 1 leader and 1 learner, raft quorum is 1. In step 4, leader automatically assumes the ReadIndex request is from local, appends ReadIndex information to local readState, and return. Therefore in this case, the learner will not get a response from leader, which leads to request timeout.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/raft/raft.go#L1047-L1049

An unit test to reproduce:
```
func TestReadOnlyOptionSafeWithLearner(t *testing.T) {
	a := newTestLearnerRaft(1, []uint64{1}, []uint64{2}, 10, 1, NewMemoryStorage())
	b := newTestLearnerRaft(2, []uint64{1}, []uint64{2}, 10, 1, NewMemoryStorage())

	nt := newNetwork(a, b)
	setRandomizedElectionTimeout(b, b.electionTimeout+1)

	for i := 0; i < b.electionTimeout; i++ {
		b.tick()
	}
	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})

	if a.state != StateLeader {
		t.Fatalf("state = %s, want %s", a.state, StateLeader)
	}

	tests := []struct {
		sm        *raft
		proposals int
		wri       uint64
		wctx      []byte
	}{
		{a, 10, 11, []byte("ctx1")},
		{b, 0, 11, []byte("ctx2")},
		{a, 10, 21, []byte("ctx3")},
		{b, 0, 21, []byte("ctx4")},
	}

	for i, tt := range tests {
		for j := 0; j < tt.proposals; j++ {
			nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgProp, Entries: []pb.Entry{{}}})
		}

		nt.send(pb.Message{From: tt.sm.id, To: tt.sm.id, Type: pb.MsgReadIndex, Entries: []pb.Entry{{Data: tt.wctx}}})

		r := tt.sm
		if len(r.readStates) == 0 {
			t.Fatalf("#%d: len(readStates) = 0, want non-zero", i)
		}
		rs := r.readStates[0]
		if rs.Index != tt.wri {
			t.Errorf("#%d: readIndex = %d, want %d", i, rs.Index, tt.wri)
		}

		if !bytes.Equal(rs.RequestCtx, tt.wctx) {
			t.Errorf("#%d: requestCtx = %v, want %v", i, rs.RequestCtx, tt.wctx)
		}
		r.readStates = nil
	}
}
```

Result:
```
--- FAIL: TestReadOnlyOptionSafeWithLearner (0.00s)
    raft_test.go:XXXX: #1: len(readStates) = 0, want non-zero
```

--- a/raft/raft.go
+++ b/raft/raft.go
@@ -1044,8 +1044,12 @@ func stepLeader(r *raft, m pb.Message) error {
 					r.send(pb.Message{To: m.From, Type: pb.MsgReadIndexResp, Index: ri, Entries: m.Entries})
 				}
 			}
-		} else {
-			r.readStates = append(r.readStates, ReadState{Index: r.raftLog.committed, RequestCtx: m.Entries[0].Data})
+		} else { // there is only one voting member (the leader) in the cluster
+			if m.From == None || m.From == r.id { // from leader itself
+				r.readStates = append(r.readStates, ReadState{Index: r.raftLog.committed, RequestCtx: m.Entries[0].Data})
+			} else { // from learner member
+				r.send(pb.Message{To: m.From, Type: pb.MsgReadIndexResp, Index: r.raftLog.committed, Entries: m.Entries})
+			}
 		}
 
 		return nil

--- a/raft/raft_test.go
+++ b/raft/raft_test.go
@@ -2416,6 +2416,57 @@ func TestReadOnlyOptionSafe(t *testing.T) {
 	}
 }
 
+func TestReadOnlyWithLearner(t *testing.T) {
+	a := newTestLearnerRaft(1, []uint64{1}, []uint64{2}, 10, 1, NewMemoryStorage())
+	b := newTestLearnerRaft(2, []uint64{1}, []uint64{2}, 10, 1, NewMemoryStorage())
+
+	nt := newNetwork(a, b)
+	setRandomizedElectionTimeout(b, b.electionTimeout+1)
+
+	for i := 0; i < b.electionTimeout; i++ {
+		b.tick()
+	}
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	if a.state != StateLeader {
+		t.Fatalf("state = %s, want %s", a.state, StateLeader)
+	}
+
+	tests := []struct {
+		sm        *raft
+		proposals int
+		wri       uint64
+		wctx      []byte
+	}{
+		{a, 10, 11, []byte("ctx1")},
+		{b, 10, 21, []byte("ctx2")},
+		{a, 10, 31, []byte("ctx3")},
+		{b, 10, 41, []byte("ctx4")},
+	}
+
+	for i, tt := range tests {
+		for j := 0; j < tt.proposals; j++ {
+			nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgProp, Entries: []pb.Entry{{}}})
+		}
+
+		nt.send(pb.Message{From: tt.sm.id, To: tt.sm.id, Type: pb.MsgReadIndex, Entries: []pb.Entry{{Data: tt.wctx}}})
+
+		r := tt.sm
+		if len(r.readStates) == 0 {
+			t.Fatalf("#%d: len(readStates) = 0, want non-zero", i)
+		}
+		rs := r.readStates[0]
+		if rs.Index != tt.wri {
+			t.Errorf("#%d: readIndex = %d, want %d", i, rs.Index, tt.wri)
+		}
+
+		if !bytes.Equal(rs.RequestCtx, tt.wctx) {
+			t.Errorf("#%d: requestCtx = %v, want %v", i, rs.RequestCtx, tt.wctx)
+		}
+		r.readStates = nil
+	}
+}
+
 func TestReadOnlyOptionLease(t *testing.T) {
 	a := newTestRaft(1, []uint64{1, 2, 3}, 10, 1, NewMemoryStorage())
 	b := newTestRaft(2, []uint64{1, 2, 3}, 10, 1, NewMemoryStorage())

